### ISSUE TITLE
dominantSpeakerChanged event not emitted if the last dominant speaker mute its microphone

### ISSUE THREAD

**Issue by SineGrain** [twilio/twilio-video.js#1585]
## dominantSpeakerChanged event not emitted if the last dominant speaker mute its microphone

 - [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**
I reproduced the issue using  the latest version of the react example app https://github.com/twilio/twilio-video-app-react

**Actual behavior:**
Scenario:
Room with 2 participants. (just to better isolate the behaviour)

Steps to reproduce with the app :

1. Both participants join the room with microphone disabled using the mute button in the PreJoinScreen
2. Particpant1 enable its microphone -> Participant2 side a dominantSpeakerChanged event is emitted with Participant1 as value.
3. Participant2 enable its microphone -> Participant1 side a dominantSpeakerChanged event is emitted with Participant2 as value
4. Participant2 mute its microphone with the ui mute button -> Participant2 side there is no dominantSpeakerChanged event emitted even if the Participant1 is the only one with an active audio stream in the room.

**Expected behavior:**

4. Participant2 mute its microphone with the ui mute button -> Participant2 side a dominantSpeakerChanged event is emitted with Participant1 as value.

This expected behaviour would allow us to inform the other participant Participant1 via data track about its new dominantSpeaker role due to known limitation of the LocalParticipant detection as suggested in this issue https://github.com/twilio/twilio-video.js/issues/604

**Software versions:**

 - [x] Browser(s): Chrome 93
 - [x] Operating System: macOs 11.3.1
 - [x] twilio-video.js: 2.15.3
 - [x] Third-party libraries (e.g., Angular, React, etc.): React

**Comment by makarandp0** [comment]
Hello @SineGrain, Thank you for reporting this issue.

I am bit confused about your repro scenario. In your repro steps after step 2, Participant 2 (Bob) sees Participant 1 (Alice) as dominant speaker, and that does not change after step 3 and 4. So at step 4 Bob already had Alice as a dominant speaker right? 
|  Step | Alice's dominantSpeaker   | Bob's dominantSpeaker |
|---|---|---|
|1. Alice and bob join with muted track   | none  | none  |
|2. Alice enables audio   | none  | Alice  |
|3. Bob enables audio   | Bob  | Alice  |
|4. Bob disables audio   | none  | Alice  |

So after Step 4, there is no change on Bob's side, so I would not expect dominantSpeakerChange event. 

Can you please clarify the steps and expectation - Also if you can share a room-sid with the repro that would be useful.

Thanks,
Makarand

**Comment by SineGrain** [comment]
Hi Makarand, 
in your table you describe the dominant speaker as a participant state, in fact in point 3 Alice and Bob has 2 different states.
We conceive and trying to achieve dominant speaker as a room state.

Also in twilio documentation you refer to dominant speaker as a room state 
https://www.twilio.com/docs/video/detecting-dominant-speaker

```
In a multi-party video application, the dominant speaker is the Participant sharing the loudest audio track in the Room.
```

Below i rewrite the table  with dominantSpeakerChanged event emission instead of state 

Step | Alice's dominantSpeaker event emitted | Bob's dominantSpeaker event emitted
-- | -- | --
|1. Alice and bob join with muted track | none | none
|2. Alice enables audio | none | Alice
|3. Bob enables audio | Bob | none
|4. Bob disables audio | none | none

From the table above if you consider the last event emitted as the Room dominant speaker we have that the last emitted participant is Bob which is not anymore in point 4 because he has muted itself.

So in point 4 after Bob muting its microphone an event emission Bob side, with Alice as value could be helpful, even if from a participant state point of view (Bob) could be perceived as a redundant event, from a room perspective this redundancy could totally make sense (Bob side will be emitted 2 time Alice)

Again as i stated in the main issue message, this expected behaviour would allow us to inform the other participant Participant1=Alice via data track about its new dominantSpeaker role, sending a message from Bob to Alice, due to known limitation of the LocalParticipant detection as suggested in this issue #604

**Comment by makarandp0** [comment]
Thank you for the explanation @SineGrain, Now I understand your ask better. You are asking for change in behavior of dominant speaker event. The SDK  treats dominant Speaker as a state per participant and not room ( you are right that documentation is should clarify this part )  

To workaround this issue, you to could use remote participant audio track [disabled](https://media.twiliocdn.com/sdk/js/video/releases/2.17.1/docs/RemoteAudioTrack.html#event:disabled__anchor) event and clear up the dominant speaker if set to that participant. 

Thanks,
Makarand

**Comment by kumail-raza** [comment]
@makarandp0 `room.on("dominantSpeakerChanged",(participant)=>{})` should emit on last user too as it does for others. Why one have to handle the same logic on different event?

**Comment by makarandp0** [comment]
@kumail-raza - Can you elaborate the scenario where you do not see the event being emitted? 
Are you referring to step 4 in @SineGrain's table - I think `none` for   `Alice's dominantSpeaker event emitted` there means an event was emitted with the dominant participant set to `null` - @SineGrain can you please confirm. `none` in your tables is bit misleading - does it mean no event emitted or does it mean an event was emitted with null as dominant speaker?

**Comment by SineGrain** [comment]
Hi @makarandp0, you're right, i updated the table, `none=no event emitted, null=event emitted with null value`, in point 4 specifically Alice side is emitted an event with null as value, in the other cases no event is emitted.

**Comment by SineGrain** [comment]
The workaround you provided @makarandp0 
`To workaround this issue, you to could use remote participant audio track disabled event and clear up the dominant speaker if set to that participant.`
could help to exclude a participant, (which can be done also using the null value emitted in step 4 probably), but don't help to know which participant is the new DominantSpeaker, in a scenario with more than 2 participants, even knowing that Bob is not anymore the DominantSpeaker, how can we know who is the new DominantSpeaker?

**Comment by makarandp0** [comment]
Hello @SineGrain - sorry, but I am bit lost on the issue here. The dominant speaker  (SP) is a per participant concept.  In a group room if two people (Alice and Bob) are both speaking simultaneously - Alice would see Bob as DS, Bob would see Alice as DS, and rest of the participants would see whoever is louder as DS. When BOB mutes (or lowers volume) they stop being DS and everybody who previously had them as DS would be notified with the event. You do not need to use data channels to communicate between participant about new DS, sdk should fire the event appropriately.

Are you seeing something not working as I described above? Maybe another repro steps would help. 

Now if you really wanted to have a single participant as as dominant speaker for the whole room- yes you would need some communication between participants to resolve the case where Alice and Bob are both talking.

The original issue that you reported was expecting Bob to get notified of dominantSpeakerChange when he mutes his Mic.  ( "Participant2 mute its microphone with the ui mute button -> Participant2 side a dominantSpeakerChanged event is emitted with Participant1 as value."  ) - which is not what the SDK does, but that is by design.


Thanks,
Makarand
